src/OpenURLServer.js

```diff
@@ -237,8 +101,142 @@ class OpenURLServer {
-    return template(data);
+    const template = ctx.cfg.getTemplate(formName);
+    ctx.body = template(data);
+  } else {
+    await next();
+  }
+}
+
+async function maybeReturnAdminData(ctx, next) {
+  const { admindata, metadata } = ctx.state;
+  if (admindata.svc?.id === 'contextObject') {
+    ctx.body = { admindata, metadata };
+  } else {
+    await next();
+  }
+}
+
+async function constructAndMaybeReturnReshareRequest(ctx, next) {
+  const { admindata, co, svcCfg, symbol } = ctx.state;
+  ctx.cfg.log('flow', 'Construct reshare request');
+  const rr = new ReshareRequest(co);
+  const rreq = rr.getRequest();
+  rreq.requestingInstitutionSymbol = symbol.includes(':') ? symbol : `RESHARE:${symbol}`;
+
+  if (svcCfg.digitalOnly) rreq.deliveryMethod = 'URL';
+
+  ctx.cfg.log('rr', JSON.stringify(rreq));
+  if (admindata.svc?.id === 'reshareRequest') {
+    ctx.body = rreq;
+  } else {
+    ctx.state.rreq = rreq;
+    await next();
+  }
+}
+
+async function postReshareRequest(ctx, next) {
+  const { admindata, metadata, npl, rreq, service } = ctx.state;
+
+  ctx.cfg.log('flow', 'Post mod-rs request');
+  // Provide a way to provoke a failure (for testing): include ctx_FAIL in the OpenURL
+  const path = get(admindata, 'ctx.FAIL') ? '/not-there' : '/rs/patronrequests';
+  const res = await service.post(path, rreq);
+  const body = await res.text();
+  ctx.cfg.log('posted', `sent request, status ${res.status}`);
+
+  if (`${res.status}`[0] !== '2') {
+    ctx.cfg.log('error', `POST error ${res.status}:`, body);
+    ctx.throw(500, 'Error encountered submitting request to mod-rs', { expose: true });
+  }
+
+  if (admindata.svc?.id === 'json') {
+    ctx.set('Content-Type', 'text/json');
+    ctx.body = {
+      status: res.status,
+      message: body,
+      contextObject: { admindata, metadata },
+      reshareRequest: rreq,
+    };
+  } else {
+    ctx.set('Content-Type', 'text/html');
+    const status = `${res.status}`;
+    const vars = { status };
+    try {
+      vars.json = JSON.parse(body);
+    } catch (e) {
+      vars.text = body;
+    };
+
+    if (!npl) {
+      await service.getPickupLocations();
+      const location = find(service.pickupLocations, x => x.code === vars.json.pickupLocationSlug);
+      if (location) vars.pickupLocationName = location.name;
+    }
+
+    const ok = (status[0] === '2');
+    const template = ctx.cfg.getTemplate(ok ? 'good' : 'bad');
+    ctx.body = template(vars);
+  }
+};
+
+class OpenURLServer {
+  constructor(cfg) {
+    this.services = {};
+
+    const serviceConfigs = cfg.getValues().services || [];
+    Object.keys(serviceConfigs).forEach(label => {
+      this.services[label] = new OkapiSession(cfg, label, serviceConfigs[label]);
+    });
+
+    // Default service
+    if (cfg.getValues().okapiUrl) this.services[''] = new OkapiSession(cfg);
+
+    const docRoot = cfg.getValues().docRoot;
+    if (!docRoot) {
+      throw new Error('No docRoot defined in configuration');
```

Comment:
Why did you get rid of HTTPError?

Reply:
Because this is a configuration error not an HTTP error, it happens before the app is even listening for HTTP requests.

Reply:
D'oh! You're right, of course.